x-pack/plugins/security_solution/server/lib/detection_engine/rule_ad_hoc_runner/api/ad_hoc_runner/route.ts

```diff
@@ -0,0 +1,439 @@
+                dataViews: dataViewsService,
+                share,
+              },
+              spaceId,
+              startedAt: startedAt.toDate(),
+              state: statePreview,
+              logger,
+              flappingSettings: DISABLE_FLAPPING_SETTINGS,
+            })) as { state: TState });
+
+            previousStartedAt = startedAt.toDate();
+            startedAt.add(parseInterval(rule.schedule.interval));
+            invocationCount--;
+          }
+        };
+
+        switch (rule.params.type) {
+          case 'query':
+            const queryAlertType = adHocRunnerRuleTypeWrapper(
+              createQueryAlertType({
+                ...ruleOptions,
+                id: QUERY_RULE_TYPE_ID,
+                name: 'Custom Query Rule',
+              })
+            );
+            await runExecutors(
+              queryAlertType.executor,
+              queryAlertType.id,
+              queryAlertType.name,
+              rule.params,
+              () => true,
+              {
+                create: alertInstanceFactoryStub,
+                alertLimit: {
+                  getValue: () => 1000,
+                  setLimitReached: () => {},
+                },
+                done: () => ({ getRecoveredAlerts: () => [] }),
+              }
+            );
+            break;
+          case 'saved_query':
+            const savedQueryAlertType = adHocRunnerRuleTypeWrapper(
+              createQueryAlertType({
+                ...ruleOptions,
+                id: SAVED_QUERY_RULE_TYPE_ID,
+                name: 'Saved Query Rule',
+              })
+            );
+            await runExecutors(
+              savedQueryAlertType.executor,
+              savedQueryAlertType.id,
+              savedQueryAlertType.name,
+              rule.params,
+              () => true,
+              {
+                create: alertInstanceFactoryStub,
+                alertLimit: {
+                  getValue: () => 1000,
+                  setLimitReached: () => {},
+                },
+                done: () => ({ getRecoveredAlerts: () => [] }),
+              }
+            );
+            break;
+          case 'threshold':
+            const thresholdAlertType = adHocRunnerRuleTypeWrapper(
+              createThresholdAlertType(ruleOptions)
+            );
+            await runExecutors(
+              thresholdAlertType.executor,
+              thresholdAlertType.id,
+              thresholdAlertType.name,
+              rule.params,
+              () => true,
+              {
+                create: alertInstanceFactoryStub,
+                alertLimit: {
+                  getValue: () => 1000,
+                  setLimitReached: () => {},
+                },
+                done: () => ({ getRecoveredAlerts: () => [] }),
+              }
+            );
+            break;
+          case 'threat_match':
+            const threatMatchAlertType = adHocRunnerRuleTypeWrapper(
```

Comment:
`threshold` rules rely on previous execution state, to prevent alerts duplication.
How is it going to work with adhoc run?
I can see the initial state here is empty, so it is possible create duplicate alerts